### ISSUE TITLE
Opening a new buffer in a floating window during WinEnter breaks diff mode on startup

### ISSUE THREAD

**Issue by dstein64** [neovim/neovim#13720]
- `nvim --version`: v0.5.0-dev+1012-gfe1ebea33
- `vim -u DEFAULTS` (version: ) behaves differently? Yes, when using Vim's popup windows instead of Neovim's floating windows, and `bufnr` to create a buffer instead of `nvim_create_buf` (using `bufnr` instead of `nvim_create_buf` doesn't fix the issue in Neovim).
- Operating system/version: Ubuntu 20.04.1
- Terminal name/version: Konsole 19.12.3
- `$TERM`: xterm-256color

### Steps to reproduce using `nvim -u NORC`

The problem occurs when the code is executed on startup.

### Steps to reproduce using `nvim -u init.vim`

Create an `init.vim` file with the following contents, which creates a new buffer, and configures an `autocmd` to show that buffer in a floating window on `WinEnter` events.

```
let bufnr = nvim_create_buf(0, 1)
let config = {
      \   'relative': 'editor',
      \   'focusable': v:false,
      \   'width': 1,
      \   'height': 1,
      \   'row': 3,
      \   'col': 3
      \ }
autocmd WinEnter * :call nvim_open_win(bufnr, v:false, config)
```

Launch `nvim` with the config file, using diff mode to compare two files. The config file is compared to itself in the example below, but the problem is more general (i.e., similar problems arise when comparing two distinct files).

```
nvim -u init.vim -d init.vim init.vim
```

### Actual behaviour

In this case, the buffer is scrolled to the bottom. Scrolling up reveals the following screen, with incorrect colors.

![diff_actual](https://user-images.githubusercontent.com/541289/104073732-730fc380-51dc-11eb-8cca-926ae813b8a4.png)

### Expected behaviour

`:diffsplit` works correctly. Here's the output of using `diffsplit`.

```
nvim -u NORC
:source init.vim
:edit init.vim
:vertical diffsplit init.vim
```

![diff_expected](https://user-images.githubusercontent.com/541289/104073741-799e3b00-51dc-11eb-80a3-aee2b950cbe3.png)

### Additional Details

This is intended as a minimal reproducible example, extracted from a similar problem I encountered when using Neovim's startup diff mode (`nvim -d`) along with [nvim-scrollview](https://github.com/dstein64/nvim-scrollview), a plugin I recently implemented.

The problem seems related to the `let bufnr = nvim_create_buf(0, 1)` line. When changing this to e.g., `let bufnr = 1`, the problem does not occur.

**Comment by zeertzjq** [comment]
Seems fixed between v0.8.0 and v0.9.0.

**Comment by dstein64** [comment]
`git bisect` shows this was fixed by be3d30cacec141bdab6d4bfb94a6fd184479de3b, introduced in PR #21829.
